python/docs/blog/posts/2023-11-06-four-pillars-of-realtime-ai.qmd

```diff
@@ -0,0 +1,115 @@
+
+RAG
+: Add information to a vector database and provide that to the LLM for richer, more precise responses.
+  For instance, load information describing the tours to answer questions like “how long is the tour” and “what visas do I need to go on the tour”.
+
+Prompt tuning
+: As you develop your application, you’ll likely want to explore variations in how you use LLMs.
+  Changes to the prompt and how the LLM is used (for instance, asking the LLM to explain the steps, as in chain-of-thought reasoning) often lead to higher quality answers.
+
+Instructions / Few-Shot Learning
+: Some behaviors may benefit from providing some few-shot learning (examples) of how we’d like the LLM to respond.
+  For instance, this may be used to manage the tone of the responses being generated, etc.
+
+Fine-Tuning
+: For some applications, creating a fine-tuned LLM for the specific questions being asked may lead to significant improvements in response quality.
+
+These learning techniques can broadly be divided into those that can be automated or scripted – such as adding new tours to the vector database – and those which require changes to the application – such as changing from a single prompt to chain-of-thought.
+While useful, this distinction between automated and manual improvements can be blurry – for instance, you could automate the collection of new training examples and fine-tuning of a new generation of model.
+
+It will be important to identify the metrics you use for quality of responses.
+While LLMs have been shown to be effective at evaluating the quality of their own output, the quality of real-time agents should be measured in terms of how helpful the overall agent is, not just the response the LLM provided within an individual action.
+User feedback is a valuable metric, especially for identifying what parts of the real-time application delight them and which parts left them confused.
+
+At the end of the day, while foundation models make it really easy to get started with an application using generative AI, you will likely find a point where you do need to look at improving the model itself.
+This shouldn’t be avoided – it means that your application has reached a point where what matters is the quality of the model responses, and you have opportunities to improve that!
+
+## Conclusion
+These four pillars are critical to creating amazing real-time AI agents:
+
+1. Watching what happens in real-time, so the agent can respond (if appropriate) and respond with awareness of what has happened.
+2. Knowing when to act by triggering actions based on the events.
+3. Acting in response to what is happening.
+4. Learning from what has happened and the results of those actions.
+
+This list should serve as a good starting point when evaluating technologies to use when building these real-time applications.
+
+What real-time applications can you envision with the rise of LLMs?
+Share your thoughts in the comments below!
```

Comment:
Yeah, we don't have a great place for comments. Thoughts on:

1. I can mention our Slack channel, but that may be a bit removed.
2. We could start creating a GitHub Discussions thread for blog posts. This would let anyone with a GitHub account post a comment.
3. Similar to three, we could just post a link to the discussions, and let people start threads for their ideas.
4. We could post a link back to whatever social we post this on (eg., link back to hacker news).

I think 2 or 3 (either a dedicated discussion thread or a category) makes sense. I'll create a placeholder for the moment (and fill in once this post is made).